### PR TITLE
feat: extract "backend" from querier cache

### PR DESCRIPTION
**Pull request by crepererum**
The backend will implement pruning policies like LRU and TTL as well as
where/how the data is stored. Having a proper interface for that
simplifies the implementation since we don't need to have one massive
`Cache` object with a super complex mechanism.

This is for #3985.

### CODE REVIEW

File: querier/src/cache_system/backend/mod.rs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+pub mod hash_map;
+
+#[cfg(test)]
+mod test_util;
+
+/// Backend to keep and manage stored entries.
+///
+/// A backend might remove entries at any point, e.g. due to memory pressure or expiration.
+pub trait Backend: Debug + Send + 'static {
```
Comment (domodwyer) [inline querier/src/cache_system/backend/mod.rs:11]: What do you think to `CacheBackend`?

That way the consumer usage is nice and descriptive:

```rust
pub struct MyServer<T> where T: CacheBackend ...
```

vs

```rust
pub struct MyServer<T> where T: Backend ...
```

File: querier/src/cache_system/backend/test_util.rs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    B: Backend<K = u8, V = String>,
+{
+    backend.set(1, String::from("a"));
+    backend.remove(&1);
+
+    assert_eq!(backend.get(&1), None);
+}
+
+/// Test setting a new value after removing it.
+fn test_readd<B>(mut backend: B)
```
Comment (domodwyer) [inline querier/src/cache_system/backend/test_util.rs:84]: typo: 
```suggestion
fn test_read<B>(mut backend: B)
```

**Review by domodwyer — APPROVED** [review]
I love composing up small traits ❤️

File: querier/src/cache_system/backend/test_util.rs (hunk 2 above)
Comment (crepererum) [reply querier/src/cache_system/backend/test_util.rs:84]: It's actually not a type. It's "re-add" :sweat_smile:

Comment (domodwyer) [reply querier/src/cache_system/backend/test_util.rs:84]: ha fair enough!

File: querier/src/cache_system/backend/mod.rs (hunk 1 above)
Comment (crepererum) [reply querier/src/cache_system/backend/mod.rs]: done

**Pull request merged**